# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘**畅销产品**，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 第一步、读取数据
导入数据分析所需要的库，并通过pandas的read_csv函数将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame并赋值给变量original_data。 

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## 第二步、评估数据
在这一部分，我们将对上一部分建立的`original_data`这个`DataFrame`所包含的数据进行评估。  
评估主要从两个方面进行，结构和内容，即整齐度和干净度。  
结构性问题指的是不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准。  
数据的内容性问题包含丢失数据、重复数据、无效数据等。

### 二-1 评估数据整齐度

In [4]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
527854,580727,22697,GREEN REGENCY TEACUP AND SAUCER,2,12/5/2011 17:17,5.79,14096.0,United Kingdom
143041,548663,37340,MULTICOLOUR SPRING FLOWER MUG,1,4/1/2011 14:44,8.29,NaN,United Kingdom
227757,556905,84596F,SMALL MARSHMALLOWS PINK BOWL,1,6/15/2011 13:05,0.42,16904.0,United Kingdom
538600,581412,23202,JUMBO BAG VINTAGE LEAF,5,12/8/2011 14:38,2.08,14415.0,United Kingdom
348924,567474,23208,LUNCH BAG VINTAGE LEAF DESIGN,4,9/20/2011 12:53,1.65,13448.0,United Kingdom
92531,544195,22174,PHOTO CUBE,12,2/17/2011 8:58,1.65,13305.0,United Kingdom
86186,543538,22666,RECIPE BOX PANTRY YELLOW DESIGN,6,2/9/2011 13:57,2.95,16143.0,United Kingdom
425984,573340,23206,LUNCH BAG APPLE DESIGN,3,10/30/2011 11:55,1.65,14159.0,United Kingdom
51985,540681,22344,PARTY PIZZA DISH PINK POLKADOT,4,1/10/2011 16:25,0.43,NaN,United Kingdom
365981,568719,21446,12 RED ROSE PEG PLACE SETTINGS,5,9/28/2011 16:19,2.46,NaN,United Kingdom


从抽样10行数据来看，符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看，每行是一次交易，每列是交易的各个变量，不存在结构性问题。

### 二-2 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


从输出结果来看，共有541909条观察值，`Description`、`CustomerID`存在缺失值。  
此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应进行格式转换。

#### 二-2-1 评估缺失数据

##### 二-2-1-1筛选`Description`缺失的行

In [6]:
original_data[original_data["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


观察到`UnitPrice`为0，进一步确认是否均为0  
找`Description`为0的同时，是否有`UnitPrice`不为0的数据

In [7]:
original_data[(original_data["Description"].isnull())&(original_data["UnitPrice"]!=0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


输出结果为空，即`Description`为0时，`UnitPrice`也为0。   
`Description`表示产品名称，`UnitPrice`表示产品单价。数据分析目标为挖掘畅销产品，二者同时缺失时，可以认为数据无法提供有效信息，因此后续可删除。

##### 二-2-1-2筛选`CustomerID`缺失的行

In [8]:
original_data[original_data["CustomerID"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


`CustomerID`表示客户编号，不是分析畅销产品的必要变量，从结果来看，有些`CustomerID`缺失的交易数据仍然有效，因此保留此变量为空的观察值。

##### 二-2-1-3评估重复数据

有三个数据具有“唯一”性：   
InvoiceNo: 发票号码。6位数，作为交易的**唯一**标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。  
StockCode: 产品代码。5位数，作为产品的**唯一**标识符。  
CustomerID: 客户编号。5位数，作为客户的**唯一**标识符。
但，实际分析可知：   
每张发票可能对应多个产品，不同交易可能包含相同产品，同一顾客可能多次交易，因此均可以重复。   
针对此数据集，无需评估重复数据。

##### 二-2-1-4评估不一致数据（同一对象有不同名称）

可能存在不同值指代同一国家的情况

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


UK、U.K.、United Kingdom 其实都是英国，但被分开统计了。  
USA 和 United States 都是美国。  
导致计数分散，影响后续分析的准确性。  
因此应该在清洗步骤对数值进行统一，只保留一个指代值。

##### 二-2-1-5评估无效或错误数据

使用`DataFrame`的`describe`方法，对数值统计信息进行快速了解

In [10]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


从输出结果看，`Quantity`和`UnitPrice`存在负数，对后续数值分析产生影响。  
因此，依次筛选出数值为负数的观察值，进一步评估其意义。

###### 二-2-1-5-1评估`Quantity`负数值

In [11]:
original_data[original_data["Quantity"]<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


观察到Quantity负数的发票订单目前可见是C开头，即交易被取消，进一步验证是否所有均是如此。  
看看是否存在Quantity为负数，但InvoiceNO不以C开头的观察值。

In [12]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


发现有一些Quantity为负的订单，不是退货订单。   
但观察到，猜测UnitPrice为0，进一步验证是否属实。  
Q为负+订单不是C+U不为0，是否还有观察值

In [13]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")&(original_data["UnitPrice"]!=0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


无数据，这说明，Quantity为负的订单，要么是C开头（被取消了）；要么是价格为0。  
均不设销售有效数据。  
在清理数据时可以删除。

###### 二-2-1-5-2评估`UnitPrice`负数值

In [14]:
original_data[original_data["UnitPrice"]<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


结果可见，为“坏账调整”，不属于实际商品交易，在数据清洗中删除。

## 第三步、清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：
- 把 `InvoiceDate` 变量的数据类型转换为日期时间
- 把 `CustomerID` 变量的数据类型转换为字符串
- 把 `Description` 变量缺失的观察值删除
- 把 `Country` 变量值 `"USA"` 替换为 `"United States"`
- 把 `Country` 变量值 `"UK"`、`"U.K."` 替换为 `"United Kingdom"`
- 把 `Quantity` 变量值为负数的观察值删除
- 把 `UnitPrice` 变量值为负数的观察值删除

为了区分清洗后的数据与原始数据，创建新变量`cleaned_data`，让它成为`original_data`复制出的副本，之后的清理步骤都在`cleaned_data`上。

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


### 三-1 把`InvoiceData`变量的数据类型转化为时间日期类型：

In [16]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[us]

### 三-2 把CustomerID 变量的数据类型转换为字符串：

In [17]:
cleaned_data['CustomerID'] = cleaned_data["CustomerID"].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: str

删除CustomerID结尾的.0（保留第一个到倒数第二个字符）

In [18]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0,-2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: str

### 三-3 把 Description 变量缺失的观察值删除:

In [19]:
cleaned_data.dropna(subset=["Description"], inplace=True)

验证是否删除：

In [20]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

### 三-4 把`Country`变量值`"USA"`替换为`"United States"`，并检查替换后`"USA"`变量值个数：

In [21]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [22]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

### 三-5 把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`，并检查替换后`"UK"`和`"U.K."`变量值个数：

In [23]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


### 三-6 把`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量值为负数的个数：

In [24]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [25]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

### 三-7 把`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量值为负数的个数：

In [26]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [27]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`e_commerce_cleaned.csv`。

看一眼老文件，发现索引无实际含义，所以后续index=False（否则默认索引保存为单独的一列）

In [28]:
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [29]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)